mycodo/controllers/controller_trigger.py

```diff
@@ -317,36 +300,66 @@ def get_method_output(self, method_id):
+        if this_controller.method_end_time:
+            # Convert time string to datetime object
+            end_time = datetime.datetime.strptime(
+                str(this_controller.method_end_time), '%Y-%m-%d %H:%M:%S.%f')
+            if now > end_time:
+                end_time_reached = True
+
+        if finished or end_time_reached:
+            # point in time is out of method range
+            with session_scope(MYCODO_DB_PATH) as db_session:
+                # Overwrite this_controller with committable connection
+                this_controller = db_session.query(Trigger).filter(
+                    Trigger.unique_id == self.unique_id).first()
+
+                if not end_time_reached and method.should_restart(now, this_controller.method_start_time):
+                    self.logger.debug("Restart")
+                    # Method has been instructed to restart
+                    this_controller.method_start_time = now
+                    # Calculate restart value
+                    setpoint, finished = method.calculate_setpoint(now, this_controller.method_start_time)
+                else:
+                    self.logger.debug("Ended")
+                    # Duration method has ended, reset method_start_time locally and in DB
+                    this_controller.method_start_time = 'Ended'
+                    this_controller.method_end_time = None
+                    this_controller.is_activated = False
+                    db_session.commit()
+
+                    self.is_activated = False
+                    self.stop_controller()
+
+                db_session.commit()
 
         if setpoint is not None:
             if setpoint > 100:
                 setpoint = 100
             elif setpoint < 0:
                 setpoint = 0
 
-        if ended:
-            with session_scope(MYCODO_DB_PATH) as db_session:
-                mod_conditional = db_session.query(Trigger)
-                mod_conditional = mod_conditional.filter(
-                    Trigger.unique_id == self.unique_id).first()
-                mod_conditional.is_activated = False
-                db_session.commit()
-            self.is_activated = False
-            self.stop_controller()
-
-        return setpoint, ended
+        return setpoint, finished
 
-    def set_output_duty_cycle(self, output_id, duty_cycle, output_channel=0):
+    def set_output_duty_cycle(self, duty_cycle):
         """ Set PWM Output duty cycle """
+        output_channel = db_retrieve_table_daemon(OutputChannel).filter(
+            OutputChannel.unique_id == self.trigger.unique_id_3).first()
+        output_channel = output_channel.channel if output_channel else 0
+        self.logger.debug("Set output duty cycle to {}".format(duty_cycle))
         self.control.output_on(
-            output_id, output_type='pwm', amount=duty_cycle, output_channel=output_channel)
+            self.trigger.unique_id_2, output_type='pwm', amount=duty_cycle, output_channel=output_channel)
```

Comment:
the db access was repeated multiple times and therefore moved into the set_output_duty_cycle method.